java/src/main/java/com/logicalclocks/hsfs/engine/HudiEngine.java

```diff
@@ -0,0 +1,293 @@
+    }
+  }
+
+  private Map<String, String> setupHudiWriteOpts(FeatureGroup featureGroup) throws IOException, FeatureStoreException {
+
+    Map<String, String> hudiArgs = new HashMap<String, String>();
+
+    hudiArgs.put(Constants.HUDI_TABLE_STORAGE_TYPE, Constants.HUDI_COPY_ON_WRITE);
+
+    Seq<String> primaryColumns = utils.getPrimaryColumns(featureGroup);
+
+    // primary key is required
+    if (primaryColumns.isEmpty()) {
+      throw new FeatureStoreException("For time travel enabled feature groups You must provide at least 1 primary key");
+    }
+
+    hudiArgs.put(Constants.HUDI_RECORD_KEY, primaryColumns.mkString(","));
+
+    // BUG: https://github.com/apache/hudi/pull/1984
+    // hudiArgs.put(Constants.HUDI_KEY_GENERATOR_OPT_KEY, Constants.HUDI_COMPLEX_KEY_GENERATOR_OPT_VAL);
+
+    Seq<String> partitionColumns = utils.getPartitionColumns(featureGroup);
+
+    if (!partitionColumns.isEmpty()) {
+      hudiArgs.put(Constants.HUDI_PARTITION_FIELD, partitionColumns.mkString(","));
+      // For precombine key take 1st particion key
+      hudiArgs.put(Constants.HUDI_PRECOMBINE_FIELD, partitionColumns.head());
+      hudiArgs.put(Constants.HUDI_HIVE_SYNC_PARTITION_FIELDS, partitionColumns.mkString(","));
+    }
+
+    String tableName = featureGroup.getName() + "_" + featureGroup.getVersion();
+    hudiArgs.put(Constants.HUDI_TABLE_NAME, tableName);
+    hudiArgs.put(Constants.HIVE_PARTITION_EXTRACTOR_CLASS_OPT_KEY,
+                Constants.DEFAULT_HIVE_PARTITION_EXTRACTOR_CLASS_OPT_VAL);
+    // Hive
+    hudiArgs.put(Constants.HUDI_HIVE_SYNC_ENABLE, "true");
+    hudiArgs.put(Constants.HUDI_HIVE_SYNC_TABLE, tableName);
+    // Gets the JDBC Connector for the project's Hive metastore
+    StorageConnectorApi storageConnectorApi = new StorageConnectorApi();
+    StorageConnector storageConnector = storageConnectorApi.getByNameAndType(featureGroup.getFeatureStore(),
+                featureGroup.getFeatureStore().getName(), StorageConnectorType.JDBC);
+    String connStr = storageConnector.getConnectionString();
+    String pw = FileUtils.readFileToString(new File("material_passwd"));
+    String jdbcUrl = connStr + "sslTrustStore=t_certificate;trustStorePassword="
+                + pw + ";sslKeyStore=k_certificate;keyStorePassword=" + pw;
+    hudiArgs.put(Constants.HUDI_HIVE_SYNC_JDBC_URL, jdbcUrl);
+    hudiArgs.put(Constants.HUDI_HIVE_SYNC_DB, featureGroup.getFeatureStore().getName());
+
+    return hudiArgs;
+  }
+
+  //hudi time time travel sql query
+  public void registerTemporaryTable(SparkSession sparkSession, FeatureGroup featureGroup, String alias,
+                                     Long startTimestamp, Long  endTimestamp) {
+
+    sparkSession.conf().set("spark.sql.hive.convertMetastoreParquet", "false");
+    sparkSession.sparkContext().hadoopConfiguration().setClass("mapreduce.input.pathFilter.class",
+        org.apache.hudi.hadoop.HoodieROTablePathFilter.class, org.apache.hadoop.fs.PathFilter.class);
```

Comment:
Can't we set these globally when we instantiate the SparkEngine?

Reply:
I think this is only needed for Hudi, for this particular function. Otherwise we need to check if FG is hudi type and set it globally in SparkEngine

Reply:
These configurations are global anyway. They will be applied globally, but only after you have read the first Hudi feature group. I'd say we are consistent and setup Spark properly when starting.